node/src/Pos/Explorer/DB.hs

```diff
@@ -104,10 +133,116 @@ prepareExplorerDB = do
             addressCoinPairs = utxoToAddressCoinPairs utxo
         putGenesisBalances addressCoinPairs
         putInitFlag
+
     -- Smooth migration for CSE-228.
     unlessM utxoSumInitializedM $ do
         putCurrentUtxoSum
 
+    -- Smooth migration for CSE-236.
+    unlessM blockEpochPagesInitializedM $ do
+        convertOldEpochBlocksFormat
+  where
+    -- | If we have persisted the 0 @Epoch@ with the old format, we need to migrate it.
+    -- Since we start syncing from the start, this will detect if the user has the old
+    -- or the new format.
+    -- Returns @True@ when the old format doesn't exist.
+    blockEpochPagesInitializedM :: MonadDBRead m => m Bool
+    blockEpochPagesInitializedM = isNothing <$> dbGet GStateDB existingOldEpochBlocks
+      where
+        existingOldEpochBlocks :: ByteString
+        existingOldEpochBlocks = previousEpochBlocksPrefix $ EpochIndex minBound
+
+    -- | This is where we convert the old format of @Epoch@ which has 21600 blocks to the
+    -- new and shiny one with @Page@s added.
+    convertOldEpochBlocksFormat :: (MonadDBRead m, MonadDB m) => m ()
+    convertOldEpochBlocksFormat =
+        runConduitRes $  epochsSource
+                      .| epochPagesConduit
+                      .| epochPagesExplorerOpConduit
+                      .| persistExplorerOpSink
+      where
+        -- 'Source' corresponding to the old @Epoch@ format that contained all
+        -- 21600 @[HeaderHash]@ in a single @Epoch@ (in production).
+        epochsSource :: (MonadDBRead m) => Source (ResourceT m) (Epoch, [HeaderHash])
+        epochsSource = dbIterSource GStateDB (Proxy @EpochsIter)
+
+        -- 'Conduit' to turn the old @Epoch@ format that contained all
+        -- 21600 @[HeaderHash]@ in a single @Epoch@ to a new format that contains
+        -- @Epoch@ *and* @Page@ as a key since there are too many @[HeaderHash]@ to
+        -- load - performance reasons.
+        epochPagesConduit
+            :: (MonadDBRead m)
+            => Conduit (Epoch, [HeaderHash]) m (Map EpochPagedBlocksKey [HeaderHash])
+        epochPagesConduit = CL.map convertToPagedMap
+          where
+            convertToPagedMap
+                :: (Epoch, [HeaderHash])
+                -> Map EpochPagedBlocksKey [HeaderHash]
+            convertToPagedMap ehh = fromListWith (++) (convertToPaged ehh)
```

Comment:
Am I understanding correctly that `(Epoch, Page)` pairs must be unique in `convertToPaged ehh`?

Reply:
`convertToPagedMap` will return a `Map` with `(Epoch, Page)` unique, if that's what you mean.

Reply:
I mean that the result of `convertToPaged ehh` should contain unique keys, and `fromListWith` actually never performs concatenation. Is that correct?

Reply:
This is correct, will fix.